File: iOS/CA2D/Rule.swift
import Foundation

struct Rule {
    let name: String
    let survive, born, conditions: Int

    init(name: String, rule: String) {
        self.name = name

        let components = rule.componentsSeparatedByString("/")
        guard components.count == 3 else {
            fatalError("Unexpected rule string")
        }

        let combine = {(result: Int, item: Character) -> Int in result + (1 << (Int(String(item))!))}

        self.survive = components[0].characters.reduce(0, combine: combine)
        self.born = components[1].characters.reduce(0, combine: combine)
        self.conditions = Int(components[2])!
    }

    static let presets = [
        Rule(name: "Banners", rule: "2367/3457/5"),
        Rule(name: "BelZhab", rule: "23/23/8"),
        Rule(name: "BelZhab Sediment", rule: "145678/23/8"),
        Rule(name: "Bloomerang", rule: "234/34678/24"),
        Rule(name: "Bombers", rule: "345/24/25"),
        Rule(name: "Brain 6", rule: "6/246/3"),
        Rule(name: "Brian's Brain", rule: "/2/3"),
        Rule(name: "Burst", rule: "0235678/3468/9"),
        Rule(name: "BurstII", rule: "235678/3468/9"),
        Rule(name: "Caterpillars", rule: "124567/378/4"),
        Rule(name: "Chenille", rule: "05678/24567/6"),
        Rule(name: "Circuit Genesis", rule: "2345/1234/8"),
        Rule(name: "Cooties", rule: "23/2/8"),
        Rule(name: "Ebb&Flow", rule: "012478/36/18"),
        Rule(name: "Ebb&Flow II", rule: "012468/37/18"),
        Rule(name: "Faders", rule: "2/2/25"),
        Rule(name: "Fireworks", rule: "2/13/21"),
        Rule(name: "Flaming Starbows", rule: "347/23/8"),
        Rule(name: "Frogs", rule: "12/34/3"),
        Rule(name: "Frozen spirals", rule: "356/23/6"),
        Rule(name: "Glisserati", rule: "035678/245678/7"),
        Rule(name: "Glissergy", rule: "035678/245678/5"),
        Rule(name: "Lava", rule: "12345/45678/8"),
        Rule(name: "Lines", rule: "012345/458/3"),
        Rule(name: "LivingOn TheEdge", rule: "345/3/6"),
        Rule(name: "Meteor Guns", rule: "01245678/3/8"),
        Rule(name: "Nova", rule: "45678/2478/25"),
        Rule(name: "OrthoGo", rule: "3/2/4"),
        Rule(name: "Prairie on fire", rule: "345/34/6"),
        Rule(name: "RainZha", rule: "2/23/8"),
        Rule(name: "Rake", rule: "3467/2678/6"),
        Rule(name: "SediMental", rule: "45678/25678/4"),
        Rule(name: "Snake", rule: "03467/25/6"),
        Rule(name: "SoftFreeze", rule: "13458/38/6"),
        Rule(name: "Spirals", rule: "2/234/5"),
        Rule(name: "Star Wars", rule: "345/2/4"),
        Rule(name: "Sticks", rule: "3456/2/6"),
        Rule(name: "Swirl", rule: "23/34/8"),
        Rule(name: "ThrillGrill", rule: "1234/34/48"),
        Rule(name: "Transers", rule: "345/26/5"),
        Rule(name: "TransersII", rule: "0345/26/6"),
        Rule(name: "Wanderers", rule: "345/34678/5"),
        Rule(name: "Worms", rule: "3467/25/6"),
        Rule(name: "Xtasy", rule: "1456/2356/16"),
    ]
}

File: Android/app/src/main/kotlin/cc/clv/android/ca2d/Rule.kt
package cc.clv.android.ca2d

data class Rule(val name: String, val rule: String) {
    companion object {
        val presets = arrayOf(
                Rule("Banners", "2367/3457/5"),
                Rule("BelZhab", "23/23/8"),
                Rule("BelZhab Sediment", "145678/23/8"),
                Rule("Bloomerang", "234/34678/24"),
                Rule("Bombers", "345/24/25"),
                Rule("Brain 6", "6/246/3"),
                Rule("Brian's Brain", "/2/3"),
                Rule("Burst", "0235678/3468/9"),
                Rule("BurstII", "235678/3468/9"),
                Rule("Caterpillars", "124567/378/4"),
                Rule("Chenille", "05678/24567/6"),
                Rule("Circuit Genesis", "2345/1234/8"),
                Rule("Cooties", "23/2/8"),
                Rule("Ebb&Flow", "012478/36/18"),
                Rule("Ebb&Flow II", "012468/37/18"),
                Rule("Faders", "2/2/25"),
                Rule("Fireworks", "2/13/21"),
                Rule("Flaming Starbows", "347/23/8"),
                Rule("Frogs", "12/34/3"),
                Rule("Frozen spirals", "356/23/6"),
                Rule("Glisserati", "035678/245678/7"),
                Rule("Glissergy", "035678/245678/5"),
                Rule("Lava", "12345/45678/8"),
                Rule("Life", "23/3/2"),
                Rule("Lines", "012345/458/3"),
                Rule("LivingOn TheEdge", "345/3/6"),
                Rule("Meteor Guns", "01245678/3/8"),
                Rule("Nova", "45678/2478/25"),
                Rule("OrthoGo", "3/2/4"),
                Rule("Prairie on fire", "345/34/6"),
                Rule("RainZha", "2/23/8"),
                Rule("Rake", "3467/2678/6"),
                Rule("SediMental", "45678/25678/4"),
                Rule("Snake", "03467/25/6"),
                Rule("SoftFreeze", "13458/38/6"),
                Rule("Spirals", "2/234/5"),
                Rule("Star Wars", "345/2/4"),
                Rule("Sticks", "3456/2/6"),
                Rule("Swirl", "23/34/8"),
                Rule("ThrillGrill", "1234/34/48"),
                Rule("Transers", "345/26/5"),
                Rule("TransersII", "0345/26/6"),
                Rule("Wanderers", "345/34678/5"),
                Rule("Worms", "3467/25/6"),
                Rule("Xtasy", "1456/2356/16")
        )
    }

    val survive: Int
    val born: Int
    val conditions: Int

    init {
        val components = rule.split("/")
        check(components.count() == 3, { "Unexpected rule string" })

        val combine = { result: Int, char: Char -> result + (1.shl(char.toString().toInt())) }
        survive = components[0].fold(0, combine)
        born = components[1].fold(0, combine)
        conditions = components[2].toInt()
    }
}
